lib/src/path_parser.cc

```diff
@@ -0,0 +1,251 @@
+                string text;
+                if (t->get_token_type() == token_type::VALUE) {
+                    // appending a number here may add a period, but we _do_ count those as path
+                    // separators, because we basically want "foo 3.0bar" to parse as a string even
+                    // though there's a number in it. The fact that we tokenize non-string values
+                    // is largely an implementation detail.
+                    auto value = tokens::get_value(t);
+
+                    // We need to split the tokens on a . so that we can get sub-paths but still preserve
+                    // the original path text when doing an insertion
+                    if (path_tokens != nullptr) {
+                        path_tokens->pop_back();
+                        for (auto&& token : split_token_on_period(t, flavor)) {
+                            path_tokens->push_back(token);
+                        }
+                    }
+                    text = value->transform_to_string();
+                } else if (t->get_token_type() == token_type::UNQUOTED_TEXT) {
+                    // We need to split the tokens on a . so that we can get sub-paths but still preserve
+                    // the original path text when doing an insertion on ConfigNodeObjects
+                    if (path_tokens != nullptr) {
+                        path_tokens->pop_back();
+                        for (auto&& token : split_token_on_period(t, flavor)) {
+                            path_tokens->push_back(token);
+                        }
+                    }
+                    text = t->token_text();
+                } else {
+                    throw config_exception("Invalid path: " + original_text +
+                                           " -- Token not allowed in path expression: " +
+                                           t->to_string() +
+                                           " (you can double-quote this token if you really want it here)");
+                }
+
+                add_path_text(elements, false, text);
+            }
+        }
+
+        path_builder builder;
+        for (element e : elements) {
+            if (e._value.length() == 0 && !e._can_be_empty) {
+                throw config_exception("Invalid path: " + original_text +
+                                               " -- path has a leading, trailing, or two adjacent '.'" +
+                " (use quoted \"\" empty string if you want an empty element)");
+            } else {
+                builder.append_key(e._value);
+            }
+        }
+
+        return builder.result();
+    }
+
+    token_list path_parser::split_token_on_period(shared_token t, config_syntax flavor) {
+        string token_text = t->token_text();
+        if (token_text == ".") {
+            return token_list { t };
+        }
+
+        vector<string> token_strings;
```

Comment:
Sorry you had to write this. I'd do something like https://github.com/puppetlabs/leatherman/blob/master/execution/src/execution.cc#L271-L275 instead. There's also http://www.boost.org/doc/libs/1_58_0/doc/html/string_algo/usage.html#idp424404576, but that creates a copy of each substring. The iterator in execution.cc iterates over token_text itself, presenting each component as a string range.

You should be able to pass string_text directly to make_split_iterator.